bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/repository/LogisticAddressRepository.kt

```diff
@@ -95,4 +98,60 @@ interface LogisticAddressRepository : JpaRepository<LogisticAddressDb, Long>, Jp
     fun joinStates(partners: Set<LogisticAddressDb>): Set<LogisticAddressDb>
+
+    @Query(
+        value =
+            """
+            SELECT la.*
+            FROM bpdm.logistic_addresses la
+            JOIN bpdm.legal_entities le ON la.legal_entity_id = le.id
+            LEFT JOIN bpdm.sites s ON s.legal_entity_id = le.id
+            WHERE
+                (
+                    CAST(:#{#searchRequest.legalName} AS text) IS NULL
+                    OR bpdm.normalize_name(le.name_value) LIKE '%' || bpdm.normalize_name(:#{#searchRequest.legalName}) || '%'
+                    OR le.name_value ~ ('.*' || REPLACE(:#{#searchRequest.legalName}, '_', '.') || '.*')
+                )
+                AND (
+                    CAST(:#{#searchRequest.bpn} AS text) IS NULL 
+                    OR le.bpn = :#{#searchRequest.bpn}
+                    OR s.bpn = :#{#searchRequest.bpn}
+                )
+                AND (
+                    CAST(:#{#searchRequest.street?.name} AS text) IS NULL
+                    OR bpdm.normalize_name(la.phy_street_name) LIKE '%' || bpdm.normalize_name(:#{#searchRequest.street?.name}) || '%'
+                    OR la.phy_street_name ~ ('.*' || REPLACE(:#{#searchRequest.street?.name}, '_', '.') || '.*')
+                )
+                AND ( 
+                    CAST(:#{#searchRequest.postalCode} AS text) IS NULL 
+                    OR la.phy_postcode = :#{#searchRequest.postalCode} 
+                )
+                AND (
+                    CAST(:#{#searchRequest.city} AS text) IS NULL
+                    OR bpdm.normalize_name(la.phy_city) LIKE '%' || bpdm.normalize_name(:#{#searchRequest.city}) || '%'
+                    OR la.phy_city ~ ('.*' || REPLACE(:#{#searchRequest.city}, '_', '.') || '.*')
+                )
+                AND (
+                    CAST(:#{#searchRequest.country} AS text) IS NULL
+                    OR la.phy_country = :#{#searchRequest.country}
+                )
+                AND (
+                       (
+                            CAST(:#{#isLegalEntity} AS boolean) = true
+                            AND le.legal_address_id = la.id
+                       )
+                    OR (
+                            CAST(:#{#isSite} AS boolean) = true
+                            AND s.main_address_id = la.id
+                       )
+                    OR (
+                            CAST(:#{#isAdditionalAddress} AS boolean) = true
+                            AND le.legal_address_id != la.id
+                            AND (s.id IS NULL OR s.main_address_id != la.id)
+                       )
+                )             
+            """,
+            nativeQuery = true
+    )
+    fun searchBusinessPartner(searchRequest: LegalEntityPropertiesSearchRequest, isLegalEntity: Boolean, isSite: Boolean, isAdditionalAddress: Boolean): Set<LogisticAddressDb>
```

Comment:
Th query definitely needs to be paginated and return a Page instead of a set. Otherwise you will never be able to create a correct pagination. In order to do that you need to do two things:

1. Add a pageable as a parameter to the repository method
2. Set the return type to Page<LogisticAddressDb>

From my experience this should be enough for JPA to create a paginated query out of your definition. However, I never tried this with such complex queries. For some hintes have a look at this discussion: https://stackoverflow.com/questions/38349930/spring-data-and-native-query-with-pagination

The third answer is what I would go for first and only adapt if it somehow does not work.